python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,758 @@
+        """
+        Triggers the caching or recaching of engine commands.
+
+        :param dict data: The data passed down from the wss client.
+        :param dict config_data: A dictionary keyed by PipelineConfiguration
+            id containing a dict that contains, at a minimum, "lookup_hash"
+            and "contents_hash" keys.
+        """
+        self.logger.debug("Caching engine commands...")
+
+        script = os.path.join(
+            os.path.dirname(__file__),
+            "scripts",
+            "cache_commands.py"
+        )
+
+        self.logger.debug("Executing script: %s" % script)
+        hash_data = dict()
+
+        for pc_id, pc_data in config_data.iteritems():
+            hash_data[pc_id] = dict(
+                lookup_hash=config_data[pc_id]["lookup_hash"],
+                contents_hash=config_data[pc_id]["contents_hash"],
+            )
+
+        args_file = self._get_arguments_file(
+            dict(
+                cache_file=self._cache_path,
+                data=data,
+                sys_path=sys.path,
+                base_configuration=self.BASE_CONFIG_URI,
+                hash_data=hash_data,
+                engine_name=self.ENGINE_NAME,
+            )
+        )
+
+        output = None
+
+        try:
+            output = subprocess.check_output([sys.executable, script, args_file])
+        except subprocess.CalledProcessError:
+            if output:
+                self.logger.error(output)
+            raise
+        else:
+            self.logger.debug(output)
+
+        self.logger.debug("Caching complete.")
+
+    def _clone_configuration(self, data):
+        """
+        Clones a pipeline configuration.
+
+        :param dict data: The payload from the client. Requires the
+            special "custom_data" key, which contains the relevant
+            path data for the cloning operation.
+        """
+        # Custom data passed in entity_type:
+        # USER_ID:NAME:LINUX_PATH:MAC_PATH:WINDOWS_PATH
+        fields = data["custom_data"].split(":")
+        user_id = int(fields[0])
+        new_name = fields[1]
+        new_path_linux = fields[2]
+        new_path_mac = fields[3]
+
+        # Note: Since the windows path may contain colons, assume all items
+        # past the 4th chunk in the command is part of the windows path.
+        new_path_windows = ":".join(fields[4:])
+        pc_entity_id = data.get("entity_id") or data["entity_ids"][0]
+
+        clone_pipeline_configuration_html(
+            self.logger,
+            self._engine.sgtk,
+            pc_entity_id,
+            user_id,
+            new_name,
+            new_path_linux,
+            new_path_mac,
+            new_path_windows,
+            sgtk.pipelineconfig_utils.is_localized(data["pc_root_path"]),
+        )
+
+    def _filter_by_project(self, actions, sw_entities, project):
+        """
+        Filters out any actions that aren't permitted for the given project
```

Comment:
I feel like this method requires a bit more documentation. If I understand correctly:
- We cache actions per context and descriptor.
- Since multiple projects could be using the same descriptor, we're hoping to be able to share a cache of actions between them.
- We pass to the launch app the special flag the allows it to enumerate all software instead of just for the current project.
- We cache that result
- Then this method here culls any actions that are not accessible for the project.

If that is the case, I think it warrants an explanation in the code, cause that is complicated (also, VERY COOL).